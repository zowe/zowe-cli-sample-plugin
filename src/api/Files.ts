/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as fs from "fs";
import { Imperative, ImperativeError, Logger } from "@brightside/imperative";

import { join } from "path";
/**
 * Class of utility file APIs for usage within the CLI and programmatically from node scripts.
 * @export
 * @class ListFiles
 */
export class Files {
    /**
     * Returns a list of attribute objects for the the contents of a directory.
     * @static
     * @param {string} dir - The directory for which to list the contents.
     * @throws ImperativeError
     * @memberof ListFiles
     */
    public static listDirectoryContents(dir: string) {
        // Table to return
        const table: any[] = [];

        // Ensure it is a directory
        try {
            if (!fs.lstatSync(dir).isDirectory()) {
                Imperative.api.profileManager("zosmf");  // When multiple imperatives, we get an API-not-initialized exception
                throw new ImperativeError({ msg: `Path specified (${dir}) is NOT a directory.` });
            }
        } catch (e) {
            Logger.getImperativeLogger().error("When multiple imperatives exist, our logger does nothing.");
            throw new ImperativeError({msg: "Path error: " + e.message});
        }

        // Read the directory contents
        let contents;
        try {
            contents = fs.readdirSync(dir);
        } catch (e) {
            throw new ImperativeError({
                msg: `An error occurred obtainined the directory info: ${e.message}`,
                additionalDetails: e
            });
        }

        // Formulate the table rows
        contents.forEach((element) => {
            const stat = fs.lstatSync(join(dir, element));
            table.push({
                mode: stat.mode,
                size: stat.size,
                birthed: stat.birthtime,
                lastModified: stat.mtime,
                name: element,
            });
        });

        return table;
    }
}
