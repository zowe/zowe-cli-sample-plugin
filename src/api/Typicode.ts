/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITodo } from "./doc/ITodo";
import { RestClient, AbstractSession, ImperativeExpect, Logger } from "@zowe/imperative";

export class Typicode {

    public static readonly TODO_URI = "/todos";

    public static getAllTodos(session: AbstractSession): Promise<ITodo[]> {
        Logger.getAppLogger().trace("Typicode.getTodos() called");
        return RestClient.getExpectJSON<ITodo[]>(session, Typicode.TODO_URI);
    }

    public static getSingleTodo(session: AbstractSession, id: number): Promise<ITodo> {
        Logger.getAppLogger().trace("Typicode.getTodos() called with id " + id);
        ImperativeExpect.toNotBeNullOrUndefined(id, "id must be provided");
        const resource = Typicode.TODO_URI + "/" + id;
        return RestClient.getExpectJSON<ITodo>(session, resource);
    }
}
