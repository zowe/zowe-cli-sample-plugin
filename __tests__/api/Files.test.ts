/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

jest.mock("fs");
import * as fs from "fs";
import { Files } from "../../src/api/Files";
const EXPECTED_LSTAT_CALLS: number = 3;
describe("Files", () => {
    it("should return attributes for a mocked directory", () => {

        // Mock lstatSync to have a mocked isDirectory function
        const mockedLstat = jest.fn((args) => {
            return {
                isDirectory: jest.fn((isDirArgs) => {
                    return true;
                }),
                size: "big!",
                mode: "000",
                birthtime: "unknown",
                mtime: "unknown",
                name: args
            };
        });
        (fs.lstatSync as any) = mockedLstat;

        // Mock readdirSync to return a mocked list
        const mockedReadDir = jest.fn((readDir) => {
            return ["file1", "directory1"];
        });
        (fs.readdirSync as any) = mockedReadDir;

        const contents = Files.listDirectoryContents("fake");
        expect(contents).toMatchSnapshot();
        expect(mockedLstat).toHaveBeenCalledTimes(EXPECTED_LSTAT_CALLS);
        expect(mockedReadDir).toHaveBeenCalledTimes(1);
    });
});
