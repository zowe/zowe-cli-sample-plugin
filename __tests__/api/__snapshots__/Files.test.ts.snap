// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files should return attributes for a mocked directory 1`] = `
Array [
  Object {
    "birthed": "unknown",
    "lastModified": "unknown",
    "mode": "000",
    "name": "file1",
    "size": "big!",
  },
  Object {
    "birthed": "unknown",
    "lastModified": "unknown",
    "mode": "000",
    "name": "directory1",
    "size": "big!",
  },
]
`;
